advisor/cmdfinder.go

```diff
@@ -0,0 +1,111 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package advisor
+
+import (
+	"fmt"
+)
+
+var commandsFinder Finder = &boltFinder{}
+
+type Command struct {
+	Snap    string
+	Command string
+}
+
+func FindCommand(command string) ([]Command, error) {
+	return commandsFinder.Find(command)
+}
+
+const (
+	minLen = 3
+	maxLen = 256
+)
+
+// based on CommandNotFound.py:similar_words.py
+func similarWords(word string) []string {
+	const alphabet = "abcdefghijklmnopqrstuvwxyz-_0123456789"
+	similar := map[string]bool{}
+
+	// deletes
+	for i := range word {
+		similar[word[:i]+word[i+1:]] = true
+	}
+	// transpose
+	for i := 0; i < len(word)-1; i++ {
+		similar[word[:i]+word[i+1:i+2]+word[i:i+1]+word[i+2:]] = true
+	}
+	// replaces
+	for i := range word {
+		for _, r := range alphabet {
+			similar[word[:i]+string(r)+word[i+1:]] = true
```

Comment:
I think it's safe to skip words starting with `-` and `_` in both 'inserts' and 'replaces'